### ISSUE TITLE
Element not defined in SSR builds

### ISSUE THREAD

**Issue by 0x6e6562** [rpearce/react-medium-image-zoom#202]
This issue is a: 

- Question / support request

## Description

When this library is included in a Gatsby site, it works fine in development.

However, when a production build is generated using SSR, the absence of the `Element` object causes the process to fail:

```
1:47:39 PM: failed Building static HTML for pages - 9.002s
1:47:39 PM: error Building static HTML failed
1:47:39 PM:   121 |   overlayBgColorStart: string.isRequired,
1:47:39 PM:   122 |   parentRef: object.isRequired,
1:47:39 PM: > 123 |   portalEl: instanceOf(Element).isRequired,
1:47:39 PM:       |                       ^
1:47:39 PM:   124 |   scrollableEl: oneOfType([object, instanceOf(Element), instanceOf(Window)]).isRequired,
1:47:39 PM:   125 |   transitionDuration: number.isRequired,
1:47:39 PM:   126 |   zoomMargin: number.isRequired
1:47:39 PM: 
1:47:39 PM:   WebpackError: ReferenceError: Element is not defined
1:47:39 PM:   
1:47:39 PM:   - UncontrolledActivated.js:123 Module../node_modules/react-medium-image-zoom/d    ist/esm/UncontrolledActivated.js
1:47:39 PM:     node_modules/react-medium-image-zoom/dist/esm/UncontrolledActivated.js:123:2    3
1:47:39 PM:   
1:47:39 PM:   - index.js:1 Module../node_modules/react-medium-image-zoom/dist/esm/index.js
1:47:39 PM:     node_modules/react-medium-image-zoom/dist/esm/index.js:1:1
```

A Gatsby-specific workaround for this is:

```js
exports.onCreateWebpackConfig = ({ stage, loaders, actions }) => {
  if (stage === "build-html") {
    actions.setWebpackConfig({
      module: {
        rules: [
          {
            test: /react-medium-image-zoom/,
            use: loaders.null(),
          },
        ],
      },
    })
  }
}
```

But when doing so, you also lose the CSS normally imported thusly:

```js
import 'react-medium-image-zoom/dist/styles.css';
```

Which breaks the zooming.

So to work around this, I added the rollup build of the CSS to the Gatsby project and imported it as if it were a local CSS file.

It seems that the offending functionality is only used in `propTypes`, so I was wondering whether there is an option in `rollup` to either

- Supply some substitute for the `Element` object for non-browser environments, or
- Publish a version of the library that strips out the `propTypes` for SSR purposes (since `propTypes` are generally only used during development)

**Comment by rpearce** [comment]
Thanks for the hawkeyes on this.

I've taken out `Element` and `Window` and am attempting to accomplish the same thing with just `propTypes.object`.

I also included [`babel-plugin-transform-react-remove-prop-types`](https://github.com/oliviertassinari/babel-plugin-transform-react-remove-prop-types#readme) with `mode: 'wrap'` to try to disable prop types in production.

Can you check it out in `4.0.0-alpha.11` when you get a chance?

**Comment by 0x6e6562** [comment]
I've just tested `4.0.0-alpha.12` which fixes the missing `Element` issue 👍 

However, there is a second issue with `HTMLElement` in `focusOptionsPolyfill()` which we previously didn't get to because the process stopped with the missing `Element`:

```
  19 |   } catch (e) {}
  20 |
> 21 |   if (HTMLElement.prototype.nativeFocus === undefined && !supportsPreventScrollOption) {
     | ^
  22 |     HTMLElement.prototype.nativeFocus = HTMLElement.prototype.focus;
  23 |
  24 |     var patchedFocus = function (args) {


  WebpackError: ReferenceError: HTMLElement is not defined

  - Main-e22c20f4.js:21 focusOptionsPolyfill
    node_modules/react-medium-image-zoom/dist/esm/Main-e22c20f4.js:21:1

  - Uncontrolled.js:11 Module../node_modules/react-medium-image-zoom/dist/esm/Uncontrolled.js
    node_modules/react-medium-image-zoom/dist/esm/Uncontrolled.js:11:21

  - index.js:1 Module../node_modules/react-medium-image-zoom/dist/esm/index.js
    node_modules/react-medium-image-zoom/dist/esm/index.js:1:1
```

**Comment by rpearce** [comment]
I just _knew_ that was going to be an issue. For this, I'll need to check if some thing(s) exists first.

I (unfortunately) don't do any SSR these days; what is the best/simplest steps to reproduce this so that I can test it and stop giving you false hope?

**Comment by rpearce** [comment]
I wonder if there's a way to automate the testing of that, too...

**Comment by 0x6e6562** [comment]
To keep things easy, I've build the simplest repro I could think of: https://github.com/0x6e6562/ssr-next

Although I originally saw the SSR issue using Gatsby, you get exactly the same behavior with `next.js`.

This repro uses `next.js` since it does SSR by default, whereas Gatsby requires an extra config file to kick off builds.

Having said this, there must be a way to invoke webpack programmatically to run SSR on an arbitrary module, I just can't think of it off the top of my head (it might be an interview question for web developers, in which case I didn't get the job).

**Comment by rpearce** [comment]
Thank you for making this repro repo! I'll use it for testing.

Regarding your interview question comment, any company that asks that is abysmal at interviewing candidates and wouldn't deserve you! I'll do some digging as a way of future-proofing this issue.

**Comment by rpearce** [comment]
Update: I haven't had time in the last 24 hours to fix this, but I'll try to get to it tonight.

**Comment by rpearce** [comment]
@0x6e6562 There were a couple of issues that SSR didn't like:
* that polyfill for Safari, which I wish I had a better way of doing
* providing `document.body` and `window` as `defaultProps` to components

They should now be resolved, and you should be able to import the CSS.

**Comment by rpearce** [comment]
If you happen to get a bare-minimum example going in nextjs or gatsby for using this project, I will :100: link to it (or them) in the README and personally use it for testing releases.

**Comment by 0x6e6562** [comment]
I can confirm that this SSR is fixed with the [simple repro](https://github.com/0x6e6562/ssr-next) and in Gatsby.

As a side note to those googling this issue, you might be seeing an issue with

```js
import 'react-medium-image-zoom/dist/styles.css';
```

working in dev but get getting excluded from production builds of Gatsby and `next.js`.

IMHO this is not related to the SSR-izibility which this particular issue addresses. Rather this is related to to the many sagas of managing CSS in various Javascript framework pipelines, which is a completely different issue.

**Comment by 0x6e6562** [comment]
@rpearce Well done for chasing this issue down, much appreciated. I've been able to remove the `loaders.null()` hack from the Gatsby SSR config (as much as anything, using a library function with the name `null()` somehow irks my sense of aesthetics).

I left the repro in a state that it demonstrates this issue is fixed and could potentially be used as a test case.

I should point out there is a lingering issue of getting the CSS to actually import in production, which is what a lot of people might well run into. But I didn't want to waste too much effort chasing down the endless task of including CSS in production builds of arbitrary JS framework, i.e. this is a systemic issue with JS pipelines IMHO.

**Comment by rpearce** [comment]
In webpack, this sort of thing is regularly solved with tools like [style-loader](https://webpack.js.org/loaders/style-loader/).

What service are you using that is giving you grief on production builds? Is it Gatsby?

**Comment by henrych4** [comment]
Just ran into the issue that css was missed in production build. Eventually my colleague found the source of error and thus we would like to share here.

In [line 34](https://github.com/rpearce/react-medium-image-zoom/blob/f3cae0b65bb840077c811e712031a8df9c166d03/package.json#L34) of package.json, `sideEffects` is set to false which excludes css file imported in production build. Removing the line or setting it as `sideEffects: ["**/*.css"]`  solves the problem.

**Comment by rpearce** [comment]
@henrych4 Wow, that is an _excellent_ discovery. I will update.

**Comment by rpearce** [comment]
@henrych4 This should be published in `v4.0.1`. Thank you!

I have added you to the all-contributors list in the README! Thanks for your contribution.

<img width="443" alt="image" src="https://user-images.githubusercontent.com/592876/72898089-581e1800-3d1b-11ea-9fe7-f553b89137f7.png">
